### PR TITLE
Userdata for EC2 and ASG support

### PR DESCRIPTION
**Pull request by clhefton**
- Merged original Userdata pull request
- Added support for asg's
- Added unit tests for ec2 and asg

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -28,4 +28,6 @@ results.xml
 .idea/
 .vscode/
 !.vscode/*.sample
+tests/data/placebo/test_ec2_userdata/*
```
Comment (kapilt) [inline .gitignore]: these should be removed, when there's some clean test data to add to this pr

File: c7n/resources/asg.py [hunk 2]
```diff
@@ -1690,6 +1691,59 @@ def __call__(self, config):
...
+            for f in as_completed(futures):
+                if f.exception():
+                    self.log.error(
+                        "Error processing userdata on instance set %s", f.exception())
+                results.extend(f.result())
+        return results
+
+    def process_instance_set(self, client, launch_configs):
+        results = []
+        for lc in launch_configs:
```
Comment (kapilt) [inline c7n/resources/asg.py]: this looks odd, you've got a launch config, and your fetching the launch config again?

**Comment by kapilt** [comment]
thanks for the pr, the filter on launch configs looks a little odd, as it doesn't need to fetch the launch configs multiple times.  its also unclear if thats really the goal versus finding asgs using problematic launch configs, which the original pr handles.

File: c7n/resources/asg.py (hunk 2 above)
Comment (clhefton) [reply c7n/resources/asg.py]: Thanks for pointing out, but isn't this simply processing a subset/batch of launch configs I already fetched? The original PR for ec2 userdata did this as well so I kept it.

Comment (kapilt) [reply c7n/resources/asg.py]: two lines below, its invoking the api per launch config to refetch the same launch config,  which doesn't make any sense.

**Comment by clhefton** [comment]
Will re-submit

**Pull request closed without merging**